src/server/services/graph/timeVaryingPathConversion.js

```diff
@@ -0,0 +1,144 @@
+ */
+async function timeVaryingPathConversion(path, conn, getEdgeConversions) {
+    // Fetch all time-varying conversions for each edge in the path
+    const edgeSegments = [];
+    for (let i = 0; i < path.length - 1; ++i) {
+        const sourceId = path[i].id;
+        const destinationId = path[i + 1].id;
+        // getEdgeConversions should return sorted array of {start_time, end_time, slope, intercept}
+        const segments = await getEdgeConversions(sourceId, destinationId, conn);
+        edgeSegments.push(segments);
+    }
+
+    // Collect all unique time boundaries
+    const boundaries = new Set();
+    edgeSegments.forEach(segments => {
+        segments.forEach(seg => {
+            boundaries.add(parsePostgresDate(seg.startTime));
+            boundaries.add(parsePostgresDate(seg.endTime));
+        });
+    });
+    const sortedBoundaries = Array.from(boundaries).sort((a, b) => a - b);
+
+    // Build combined segments
+    const results = [];
+    for (let i = 0; i < sortedBoundaries.length - 1; ++i) {
+        const startTime = sortedBoundaries[i];
+        const endTime = sortedBoundaries[i + 1];
+        let valid = true;
+        let slope = 1;
+        let intercept = 0;
+
+        // For each edge, find the segment covering this time range
+        for (const [edgeIndex, segments] of edgeSegments.entries()) {
+            // Try to find segment in forward direction
+            let seg = segments.find(s =>
+                parsePostgresDate(s.startTime) <= startTime &&
+                parsePostgresDate(s.endTime) >= endTime &&
+                s.sourceId === path[edgeIndex].id &&
+                s.destinationId === path[edgeIndex + 1].id
+            );
+
+            let invert = false;
+
+            if (!seg) {
+                // Try to find segment in reverse direction
+                seg = segments.find(s =>
+                    parsePostgresDate(s.startTime) <= startTime &&
+                    parsePostgresDate(s.endTime) >= endTime &&
+                    s.sourceId === path[edgeIndex + 1].id &&
+                    s.destinationId === path[edgeIndex].id
+                );
+                if (seg) {
+                    // Check bidirectional in conversions table
+                    const bidirectional = await isBidirectional(s.sourceId, s.destinationId, conn);
+                    if (bidirectional) invert = true;
+                    else seg = null; // Not bidirectional, can't use
+                }
+            }
+
+            if (!seg) {
+                valid = false;
+                break;
+            }
+
+            let segSlope = seg.slope;
+            let segIntercept = seg.intercept;
+            if (invert) {
+                [segSlope, segIntercept] = invertConversion(segSlope, segIntercept);
+            }
+
+            [slope, intercept] = updatedConversion(slope, intercept, segSlope, segIntercept);
+        }
+        if (valid) {
```

Comment:
Should not valid be an error or some sort? Also, can it happen if the segments are correct?